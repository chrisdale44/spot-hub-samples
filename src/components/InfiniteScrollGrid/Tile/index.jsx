"use client";

import Image from "next/image";
import PropTypes from "prop-types";
import React from "react";
import { MdImageNotSupported } from "@/icons/";

const Tile = ({ item, onClick }) => {
  const { name, thumb } = item;
  return (
    <div className="relative flex flex-[1_1_200px] aspect-square min-w-[140px] lg:min-w-[180px] max-w-[225px] cursor-pointer overflow-hidden bg-[white] p-0">
      {name ? (
        <a
          className="absolute inset-x-0 top-0 z-[1] m-0 bg-[rgba(44,44,44,0.7)] p-1 text-[white] text-xs whitespace-nowrap overflow-hidden truncate"
          onClick={() => onClick(item)}
          data-testid="tile-name"
        >
          {name}
        </a>
      ) : null}
      {thumb ? (
        <Image
          className="inset-0"
          src={thumb}
          fill={true}
          alt={name}
          loading="lazy"
          decoding="async"
          onClick={() => onClick(item)}
          unoptimized
          data-testid="tile-image"
        />
      ) : (
        <div
          className="text-[#777] text-sm h-full w-full flex flex-col items-center justify-center bg-[#f9f9f9] text-center"
          onClick={() => onClick(item)}
          data-testid="no-image"
        >
          <MdImageNotSupported className="w-8 h-8 text-[#777]" />
          No image
          <br />
          available
        </div>
      )}
    </div>
  );
};

Tile.propTypes = {
  item: PropTypes.object.isRequired,
};

export default Tile;
