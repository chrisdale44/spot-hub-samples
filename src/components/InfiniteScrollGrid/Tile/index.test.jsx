import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import { describe, it, expect, vi } from "vitest";
import Tile from ".";
import { MdImageNotSupported } from "@/icons/";
import Image from "next/image";

vi.mock("next/image", () => ({
  default: vi.fn(({ fill, unoptimized, ...props }) => (
    <img
      {...props}
      data-test-fill={fill ? "true" : "false"}
      data-test-unoptimized={unoptimized ? "true" : "false"}
    />
  )),
}));

vi.mock("@/icons/", () => ({
  MdImageNotSupported: () => <span data-testid="no-image-icon" />,
}));

describe("COMPONENT: Tile", () => {
  const mockOnClick = vi.fn();
  const baseItem = {
    name: "Test Item",
    thumb: "test-thumb.jpg",
  };

  beforeEach(() => {
    vi.clearAllMocks();
  });

  it("renders correctly with name and thumb", () => {
    render(<Tile item={baseItem} onClick={mockOnClick} />);

    expect(screen.getByTestId("tile-name")).toBeInTheDocument();
    expect(screen.getByTestId("tile-image")).toHaveAttribute(
      "src",
      "test-thumb.jpg"
    );
    expect(screen.getByTestId("tile-image")).toHaveAttribute(
      "alt",
      "Test Item"
    );
  });

  it("calls onClick when clicked", () => {
    render(<Tile item={baseItem} onClick={mockOnClick} />);

    fireEvent.click(screen.getByTestId("tile-image"));
    expect(mockOnClick).toHaveBeenCalledTimes(1);
    expect(mockOnClick).toHaveBeenCalledWith(baseItem);
  });

  it("calls onClick when name link is clicked", () => {
    render(<Tile item={baseItem} onClick={mockOnClick} />);

    fireEvent.click(screen.getByTestId("tile-name"));
    expect(mockOnClick).toHaveBeenCalledTimes(1);
  });

  it("renders without name when not provided", () => {
    const itemWithoutName = { ...baseItem, name: undefined };
    render(<Tile item={itemWithoutName} onClick={mockOnClick} />);

    expect(screen.queryByText("Test Item")).not.toBeInTheDocument();
    expect(screen.queryByTestId("tile-image")).toBeInTheDocument();
  });

  describe("when thumb is not provided", () => {
    it("renders a fallback", () => {
      const itemWithoutThumb = { ...baseItem, thumb: undefined };
      render(<Tile item={itemWithoutThumb} onClick={mockOnClick} />);

      expect(screen.queryByTestId("no-image-icon")).toBeInTheDocument();
      expect(screen.queryByTestId("no-image")).toBeInTheDocument();
      expect(screen.queryByTestId("tile-image")).not.toBeInTheDocument();
    });

    it("calls onClick when fallback is clicked", () => {
      const itemWithoutThumb = { ...baseItem, thumb: undefined };
      render(<Tile item={itemWithoutThumb} onClick={mockOnClick} />);

      fireEvent.click(screen.queryByTestId("no-image"));
      expect(mockOnClick).toHaveBeenCalledTimes(1);
    });
  });

  it("matches snapshot with thumb", () => {
    const { container } = render(
      <Tile item={baseItem} onClick={mockOnClick} />
    );
    expect(container).toMatchSnapshot();
  });

  it("matches snapshot without thumb", () => {
    const itemWithoutThumb = { ...baseItem, thumb: undefined };
    const { container } = render(
      <Tile item={itemWithoutThumb} onClick={mockOnClick} />
    );
    expect(container).toMatchSnapshot();
  });
});
