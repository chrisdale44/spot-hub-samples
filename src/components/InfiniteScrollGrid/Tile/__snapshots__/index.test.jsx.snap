// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`COMPONENT: Tile > matches snapshot with thumb 1`] = `
<div>
  <div
    class="relative flex flex-[1_1_200px] aspect-square min-w-[140px] lg:min-w-[180px] max-w-[225px] cursor-pointer overflow-hidden bg-[white] p-0"
  >
    <a
      class="absolute inset-x-0 top-0 z-[1] m-0 bg-[rgba(44,44,44,0.7)] p-1 text-[white] text-xs whitespace-nowrap overflow-hidden truncate"
      data-testid="tile-name"
    >
      Test Item
    </a>
    <img
      alt="Test Item"
      class="inset-0"
      data-test-fill="true"
      data-test-unoptimized="true"
      data-testid="tile-image"
      decoding="async"
      loading="lazy"
      src="test-thumb.jpg"
    />
  </div>
</div>
`;

exports[`COMPONENT: Tile > matches snapshot without thumb 1`] = `
<div>
  <div
    class="relative flex flex-[1_1_200px] aspect-square min-w-[140px] lg:min-w-[180px] max-w-[225px] cursor-pointer overflow-hidden bg-[white] p-0"
  >
    <a
      class="absolute inset-x-0 top-0 z-[1] m-0 bg-[rgba(44,44,44,0.7)] p-1 text-[white] text-xs whitespace-nowrap overflow-hidden truncate"
      data-testid="tile-name"
    >
      Test Item
    </a>
    <div
      class="text-[#777] text-sm h-full w-full flex flex-col items-center justify-center bg-[#f9f9f9] text-center"
      data-testid="no-image"
    >
      <span
        data-testid="no-image-icon"
      />
      No image
      <br />
      available
    </div>
  </div>
</div>
`;

exports[`Tile Component > matches snapshot with thumb 1`] = `
<div>
  <div
    class="relative flex flex-[1_1_200px] aspect-square min-w-[140px] lg:min-w-[180px] max-w-[225px] cursor-pointer overflow-hidden bg-[white] p-0"
  >
    <a
      class="absolute inset-x-0 top-0 z-[1] m-0 bg-[rgba(44,44,44,0.7)] p-1 text-[white] text-xs whitespace-nowrap overflow-hidden truncate"
    >
      Test Item
    </a>
    <img
      alt="Test Item"
      class="inset-0"
      decoding="async"
      loading="lazy"
      src="test-thumb.jpg"
    />
  </div>
</div>
`;

exports[`Tile Component > matches snapshot without thumb 1`] = `
<div>
  <div
    class="relative flex flex-[1_1_200px] aspect-square min-w-[140px] lg:min-w-[180px] max-w-[225px] cursor-pointer overflow-hidden bg-[white] p-0"
  >
    <a
      class="absolute inset-x-0 top-0 z-[1] m-0 bg-[rgba(44,44,44,0.7)] p-1 text-[white] text-xs whitespace-nowrap overflow-hidden truncate"
    >
      Test Item
    </a>
    <div
      class="text-[#777] text-sm h-full w-full flex flex-col items-center justify-center bg-[#f9f9f9] text-center"
    >
      <span
        data-testid="no-image-icon"
      />
      No image
      <br />
      available
    </div>
  </div>
</div>
`;
